exercises/circular-buffer/example.rs

```diff
@@ -55,6 +55,11 @@ impl<T: Default + Clone> CircularBuffer<T> {
     pub fn clear(&mut self) {
         self.start = 0;
         self.end = 0;
+
+        // Clear any values in the buffer
+        for element in self.buffer.iter_mut() {
+            std::mem::take(element);
+        }
```

Comment:
Why not just [delegate the method](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.clear): 
```suggestion
        self.buffer.clear();
```

Reply:
We can't do that. If you look at how the `new()`, all of the code relies on the buffer always being populated and being `size + 1`:

```rust
pub fn new(size: usize) -> CircularBuffer<T> {
    CircularBuffer {
        buffer: vec![T::default(); size + 1],
        size: size + 1,
        start: 0,
        end: 0,
    }
}
```

Doing `mem::take()` on each element simply returns them all back to `T::default()`, but leaves the vector layout intact, while doing `self.buffer.clear()` removes all the elements.

Now, arguably, having a `T::default()` in each slot isn't very good practice; in my own solution I went with `Option<T>` instead of `T: Default` so I could store `None` values for unused slots instead of whatever `T`'s default is. This also let me avoid the `T: Default` bound. But now I'm just pontificating. 😉

Reply:
Right, fair enough. I'd forgotten that the implementation depended on that assumption.